lte/gateway/python/magma/kernsnoopd/ebpf/byte_count.bpf.c

```diff
@@ -23,28 +24,48 @@ BPF_HASH(dest_counters, struct key_t);
 
 // Attach kprobe for the `tcp_sendmsg` syscall
 int kprobe__tcp_sendmsg(struct pt_regs *ctx, struct sock *sk, struct msghdr *msg, size_t size) {
-  u16 dport = 0, family = sk->__sk_common.skc_family;
+  u16 family = sk->sk_family;
+
+  // both IPv4 and IPv6
+  if (family == AF_INET || family == AF_INET6) {
+    struct key_t key = {};
 
-  // only IPv4
-  if (family == AF_INET) {
-    struct key_t key;
-    key.daddr = sk->__sk_common.skc_daddr;
+    // read destination IP address and port
+    if (family == AF_INET) {
+      key.daddr = sk->sk_daddr;
+    }
+    else if (family == AF_INET6) {
+      bpf_probe_read(&key.daddr, sizeof(key.daddr), &sk->sk_v6_daddr.s6_addr32);
+    }
+    u16 dport = sk->sk_dport;
+    key.dport = ntohs(dport);
+
+    // ignore packets destined for localhost unless dport is PROXY_PORT
 
-    // ignore packets destined for localhost
     // IPv4 127.0.0.1 == 16777343, 0.0.0.0 == 0
-    if ((key.daddr == 16777343 || key.daddr == 0)) {
-      return 0;
+    if (family == AF_INET && key.dport != {{PROXY_PORT}}) {
+      if (key.daddr == 0x100007F || key.daddr == 0) {
+        return 0;
+      }
+    }
+    // IPv6 localhost has embedded IPv4 in case of control_proxy
+    // uint128 is split into two u64 and high is FFFF:"127.0.0.1"
+    else if (family == AF_INET6 && key.dport != {{PROXY_PORT}}) {
+      uint64_t low = (uint64_t)key.daddr;
+      uint64_t high = (key.daddr >> 64);
+      if (low == 0 && high == 0x100007FFFFF0000) {
```

Comment:
it would be easier/faster to define constant unsigned char array and use memcmp() here.

Reply:
Looked into it for a while. This would be the required byte string to compare with:
```
"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xFF\xFF\x7F\x00\x00\x01"
```

However, `memcmp` isn't allowed in eBPF. So if you use it, the program will compile but fail to load as the verifier will reject it. `__builtin_memcmp`, which should unroll at compile time, does not work because of a [clang bug](https://bugs.llvm.org/show_bug.cgi?id=26218). Now that bounded loops are supported in eBPF, one could write a custom memcmp(), but at that point, the two-part comparison above seems better to me.